### PR TITLE
Feature fnull

### LINKED ISSUE #1235: fnull
**Is your feature request related to a problem? Please describe.**
fnull is in a function that takes a function as an argument and a number of additional arguments, and returns a function that just calls the original function given. The magic of fnull is that if any of the arguments to the function that it returns are null or undefined, then the original “default” argument is used instead.

**Describe the solution you'd like**
```js
const multiply = (a, b) => a * b;
const safeMultiply = fnull(multiply, [1, 1]);

safeMultiply(2, null); //=> 2
```

**Describe alternatives you've considered**

https://www.linkedin.com/pulse/function-currying-default-params-javascript-vladim%C3%ADr-gorej/

**Additional context**

### PR DESCRIPTION
**Pull request by sidag95**
fnull is in a function that takes a function as an argument and a number of additional arguments, and returns a function that just calls the original function given. The magic of fnull is that if any of the arguments to the function that it returns are null or undefined, then the original “default” argument is used instead

Closes #1235

### CODE REVIEW

**Comment by char0n** [comment]
@sidag95 thank you for your contribution! I'll make a review during today. I was on skiing trip last 4 days. Sorry for CR delay!

**Comment by sidag95** [comment]
> @sidag95 thank you for your contribution! I'll make a review during today. I was on skiing trip last 4 days. Sorry for CR delay!

No worries!

File: src/fNull.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Returns a function which is called with the given arguments. If any of the given arguments are null or undefined,
+ * the corresponding default value for that argument is used instead to call the original function.
+ *
+ * @func fNull
+ * @memberOf RA
+ * @category Function
+ * @sig function, [defaultArgs] -> function -> function(args or defaultArgs)
+ * @param {Function} function to be executed
+ * @param {Array|undefined} default arguments to call the function with
```
Comment (char0n) [inline src/fNull.js]:
```suggestion
 * @param {Array} defaults default arguments
```

File: src/fNull.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * @return {Function} function to be called
```
Comment (char0n) [inline src/fNull.js]:
```suggestion
 * @return {Function} function will apply provided arguments or default ones
```

File: src/fNull.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * @example
+ *
+ * RA.fNull(); // => fn // => undefined
```
Comment (char0n) [inline src/fNull.js]:
```suggestion
```

File: src/fNull.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * RA.fNull(fn); // => fn // => fn(args) //no default arguments
```
Comment (char0n) [inline src/fNull.js]:
```suggestion
```

File: src/fNull.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * RA.fNull(fn, [a]) // => fn // => fn(args) first argument with default param as a
```
Comment (char0n) [inline src/fNull.js]: Here we're trying to demonstrate the functionality of the function.

```suggestion
 const addDefaults = RA.fnull((a, b) => a + b, [4, 5])
 
 addDefaults(1, 2); // => 3
 addDefaults(null, 2); // => 6
 addDefaults(2, null); // => 7
 addDefaults(undefined, undefined); // => 9 
```

File: src/fNull.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+const fNull = curryN(2, (originalFunction, defaultArgs) => {
```
Comment (char0n) [inline src/fNull.js]: Couple of notes here:

- Function needs to be of arity 2 (satisfied)
- Function needs to return function of the same arity as original function (not satisfied)
- Function must be total (not satisfied)
- No parts of the arguments nor return value must be checked (not satisfied)

Here is my proposal of the function. 

```js
const fnull = curry((fn, defaults) => {
  return curryN(fn.length, (...args) => {
    const argsWithDefaults = mapIndexed(
      (val, idx) => defaultWhen(isNil, defaults[idx], val),
      args
    );

    return apply(fn, argsWithDefaults);
  });
});
```
I've tried to use functions that we already have in R or RA.

File: src/index.js [hunk 7]
```diff
@@ -112,6 +112,7 @@ export { default as async } from './async';
 export { default as anyP } from './anyP';
 export { default as firstP } from './anyP'; // alias of anyP
 export { default as lastP } from './lastP';
+export { default as fNull } from './fNull';
```
Comment (char0n) [inline src/index.js]:
```suggestion
export { default as fnull } from './fnull';
```

File: types/index.d.ts [hunk 8]
```diff
@@ -1140,6 +1140,12 @@ declare namespace RamdaAdjunct {
         argsPass<T>(combiningPredicate: (fn: (a: T) => boolean) => (list: T[]) => boolean, predicates: Pred[]): Pred;
         argsPass<T>(combiningPredicate: (fn: (a: T) => boolean) => (list: T[]) => boolean): (predicates: Pred[]) => Pred;
 
+        /**
+         * Takes a function as an argument and a number of additional arguments, and returns a function that just calls the original function given.
```
Comment (char0n) [inline types/index.d.ts]: Please align the description of the function with the one defined in https://github.com/char0n/ramda-adjunct/pull/1321/files#diff-00827205c5757eed238b3d5634928ae5R8

File: types/index.d.ts [hunk 9]
```diff
@@ -1140,6 +1140,12 @@ declare namespace RamdaAdjunct {
...
+         * If any of the arguments to the function that it returns are null or undefined, then the original default argument is used instead.
+         */
+         fNull(fn: Function, defaultArgs: [any]): Function;
```
Comment (char0n) [inline types/index.d.ts]:
```suggestion
         fnull(fn: Function, defaults: any[]): Function;
         fnull(fn: Function): (defaults: any[]) => Function;  
```

Comment (char0n) [reply types/index.d.ts]: We're missing the tests for these typings: please have a look at model PR which shows what needs to be done: https://github.com/char0n/ramda-adjunct/pull/1287/files

File: test/fNull.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
+import { assert } from 'chai';
+
+import * as RA from '../src';
+
+describe('fNull', function() {
```
Comment (char0n) [inline test/fNull.js]: We'll get into reviewing the tests when we have more correct implementation.

**Comment by sidag95** [comment]
@char0n Thank you so much for such a detailed review. I will work on the suggested changes and will keep asking questions as and when they arrive.

> Couple of notes here:
Function needs to be of arity 2 (satisfied)
Function needs to return function of the same arity as original function (not satisfied)
Function must be total (not satisfied)
No parts of the arguments nor return value must be checked (not satisfied)

One immediate question in my mind is how to know about these notes? As in, how do we define these constraints for this function?

**Comment by char0n** [comment]
> @char0n Thank you so much for such a detailed review. I will work on the suggested changes and will keep asking questions as and when necessary.

No problem

> One immediate question in my mind is how to know about these notes? As in, how do we define these constraints for this function?

The model implementation I provided satisfies all those requirements. Try to map those requirements to the implementation. 

```js
const fnull = curry((fn, defaults) => {
  return curryN(fn.length, (...args) => {
    const argsWithDefaults = mapIndexed(
      (val, idx) => defaultWhen(isNil, defaults[idx], val),
      args
    );

    return apply(fn, argsWithDefaults);
  });
});
```

- Function needs to be of arity 2 (satisfied) - **line 1**
- Function needs to return function of the same arity as original function (not satisfied) - **line 2**
- Function must be total (not satisfied) - **line 8** (returns only what original fn returns)
- No parts of the arguments nor return value must be checked (not satisfied) - we're not checking anything, we assume user provides what we expect.

File: src/fNull.js (hunk 6 above)
Comment (sidag95) [reply src/fNull.js]: Aah got it after taking my time to understand this code. 
The artiy of the function is important to enable currying for each argument. 
Also, the function did not allow currying at all, the way I had implemented it. 

Learnt something new today (total functions)

Comment (char0n) [reply src/fNull.js]: Right. This is a good place to study the FP jargon: https://github.com/hemanth/functional-programming-jargon

**Comment by sidag95** [comment]
I have made the required changes and have updated the tests as well. Please have a look whenever you get time. Thanks!

File: src/fNull.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import isArray from './isArray';
+
+/**
+ * Returns a function which is called with the given arguments. If any of the given arguments are null or undefined,
+ * the corresponding default value for that argument is used instead to call the original function.
+ *
+ * @func fNull
+ * @memberOf RA
+ * @category Function
+ * @sig function, [defaultArgs] -> function -> function(args or defaultArgs)
```
[the remark this answers is not in the document]
Comment (char0n) [reply src/fNull.js]: Next time you can directly use github suggestion feature to commit my suggested change to your PR: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

File: test/fnull.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+        const value = safeAdd(1, undefined);
+
+        assert.equal(value, 2);
+      }
+    );
+  });
+
+  it('should curry', async function() {
+    const add = (a, b) => a + b;
```
Comment (char0n) [inline test/fnull.js]: I suggest to simply it into following form. Also the function doesn't need to be async. In this form IMHO it is more apparent what we're testing here.
```js
it('should be curried', function() {
   const add = (a, b) => a + b;
   
   assert.strictEqual(RA.fnull(add, [1, 1]), 2);
   assert.strictEqual(RA.fnull(add)([1, 1]), 2);
});
```

File: test/fnull.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
+import { assert } from 'chai';
+
+import * as RA from '../src';
+
```
Comment (char0n) [inline test/fnull.js]:
```js
import * as R from 'ramda';
```

File: test/fnull.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+describe('fnull', function() {
+  context('given function and default arguments', function() {
+    specify(
```
Comment (char0n) [inline test/fnull.js]: What would you say if we merged two following tests (except currying test) into one clear block of assertions?
```js
context('given function and default arguments', function() {
   specify('should use argument or appropriate default argument', function() {
      assert.strictEqual(RA.fnull(R.add, [1, 1])(1, 2), 3);
      assert.strictEqual(RA.fnull(R.add, [1, 1])(null, 2), 3);
      assert.strictEqual(RA.fnull(R.add, [1, 1])(null, undefined), 2);
      ....what ever other assertion combinations you can think of using `null` or `undefined` or not nil value...
   });
});
```

File: test/fnull.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+        const safeAdd = RA.fnull(add, [1, 1]);
+
+        const value = safeAdd(1, undefined);
+
+        assert.equal(value, 2);
+      }
+    );
+  });
+
```
Comment (char0n) [inline test/fnull.js]: Also test for returning curried function would be nice
```js
it('should return curried function', function() {
  const fn = RA.fnull(R.add, [1, 1]);

  assert.lengthOf(fn, 2);
  assert.strictEqual(fn(1, 1), 2);
  assert.strictEqual(fn(1)(1), 2);
});

File: src/fNull.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (sidag95) [reply src/fNull.js]: Hey this is cool! Will keep in mind

**Comment by sidag95** [comment]
I have made the changes, please have a look whenever you get time. Thanks!

**Comment by sidag95** [comment]
Will check the error and will let you know

**Comment by sidag95** [comment]
Seems like the browser is crashing for a couple of tests
>ok 1097 Firefox 65.0 - [0 ms] - zipObjWith given key/value lists of unequal length should truncate to the length of the shortest list
not ok 1098 Safari - [undefined ms] - error
    ---
        message: >
            Error: Browser failed to connect within 30s. testem.js not loaded?

        browser log: |
            [object Object]
    ...
ok 1099 Chrome 82.0 - [0 ms] - allEqual should support placeholder to specify "gaps"


> ok 2230 Chrome 80.0 - [0 ms] - allP should have fail-fast behavior
ok 2231 Chrome 80.0 - [0 ms] - allP should support placeholder to specify "gaps"
not ok 2232 Headless Chrome - [undefined ms] - error
    ---
        message: >
            Error: Browser exited unexpectedly
            Non-zero exit code: null
            Stderr:
             [0303/214346.373150:ERROR:xattr.cc(63)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/7f/k77dk5055h117lhyl5nw5pbh0000gn/T/: Operation not permitted (1)
            [0303/214346.375707:ERROR:file_io.cc(89)] ReadExactly: expected 8, observed 0
            [0303/214346.378644:ERROR:xattr.cc(63)] setxattr org.chromium.crashpad.database.initialized on file /var/folders/7f/k77dk5055h117lhyl5nw5pbh0000gn/T/: Operation not permitted (1)

            DevTools listening on ws://127.0.0.1:53069/devtools/browser/5df4d669-5a79-4f8f-bcdb-399da1550e40
            [0303/214346.706530:WARNING:gpu_process_host.cc(1165)] The GPU process has crashed 1 time(s)
            [0303/214346.983753:WARNING:gpu_process_host.cc(934)] Reinitialized the GPU process after a crash. The reported initialization time was 0 ms
            [0303/214346.984486:WARNING:gpu_process_host.cc(1165)] The GPU process has crashed 2 time(s)
            [0303/214347.229596:WARNING:gpu_process_host.cc(934)] Reinitialized the GPU process after a crash. The reported initialization time was 0 ms
            [0303/214347.230288:WARNING:gpu_process_host.cc(1165)] The GPU process has crashed 3 time(s)
            [0303/214347.230316:FATAL:gpu_data_manager_impl_private.cc(1034)] The display compositor is frequently crashing. Goodbye.


        browser log: |
            [object Object]
            [object Object]
            [object Object]
    ...
ok 2233 Chrome 80.0 - [0 ms] - allEqual should support placeholder to specify "gaps"


I am not sure how to debug these. How should I proceed?

File: types/test/fnull.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import * as RA from 'ramda-adjunct';
+
+const fn = () => null;
```
Comment (char0n) [inline types/test/fnull.ts]: Let's define this function properly as a function receiving one argument that is being returned.
```js
const fn = v => v
```

File: types/index.d.ts [hunk 17]
```diff
@@ -1140,6 +1140,13 @@ declare namespace RamdaAdjunct {
         argsPass<T>(combiningPredicate: (fn: (a: T) => boolean) => (list: T[]) => boolean, predicates: Pred[]): Pred;
         argsPass<T>(combiningPredicate: (fn: (a: T) => boolean) => (list: T[]) => boolean): (predicates: Pred[]) => Pred;
 
+        /**
+         * Returns a function which is called with the given arguments. If any of the given arguments are null or undefined,
+         * the corresponding default value for that argument is used instead
```
Comment (char0n) [inline types/index.d.ts]:
```suggestion
         * the corresponding default value for that argument is used instead.
```

**Comment by sidag95** [comment]
I will fix those the last few issues and push an update soon, thanks!

**Comment by sidag95** [comment]
@char0n I have made all the changes, please have a look, thanks!

**Comment by sidag95** [comment]
Have reverted webpack to 4.4.2 (same as master)

File: package-lock.json [hunk 18]
```diff
@@ -5383,14 +5383,14 @@
         "json-stable-stringify": "^1.0.1",
         "strip-json-comments": "^2.0.1",
         "tslint": "5.14.0",
-        "typescript": "^3.9.0-dev.20200229",
+        "typescript": "=3.9.0-dev.20200305",
```
Comment (char0n) [inline package-lock.json]:
```suggestion
        "typescript": "^3.9.0-dev.20200229",
```

File: package-lock.json [hunk 19]
```diff
@@ -5388,9 +5388,9 @@
       },
       "dependencies": {
         "typescript": {
-          "version": "3.9.0-dev.20200229",
-          "resolved": "https://registry.npmjs.org/typescript/-/typescript-3.9.0-dev.20200229.tgz",
-          "integrity": "sha512-DtSLzxoiUir0qRc3+JJBxiAe6NvTEM3uDxnPxVWJU6sRDhUi8Ssx6DBjGWCZAQJlLk5A+jk2ptf3JvvZrQlLNQ==",
+          "version": "3.9.0-dev.20200305",
```
Comment (char0n) [inline package-lock.json]:
```suggestion
         "version": "3.9.0-dev.20200229",
```

File: package-lock.json [hunk 20]
```diff
@@ -5389,8 +5389,8 @@
       "dependencies": {
         "typescript": {
           "version": "3.9.0-dev.20200229",
-          "resolved": "https://registry.npmjs.org/typescript/-/typescript-3.9.0-dev.20200229.tgz",
-          "integrity": "sha512-DtSLzxoiUir0qRc3+JJBxiAe6NvTEM3uDxnPxVWJU6sRDhUi8Ssx6DBjGWCZAQJlLk5A+jk2ptf3JvvZrQlLNQ==",
+          "resolved": "https://registry.npmjs.org/typescript/-/typescript-3.9.0-dev.20200305.tgz",
```
Comment (char0n) [inline package-lock.json]:
```suggestion
          "resolved": "https://registry.npmjs.org/typescript/-/typescript-3.9.0-dev.20200229.tgz",
```

File: package-lock.json [hunk 21]
```diff
@@ -5389,8 +5389,8 @@
...
+          "integrity": "sha512-O24m5gu1sfYsDWIgWcA2TI2+8Ou1whbjSe9OxSPvyelJB/Vv1SeOo3IGkoudIuvt3SM/Nf34+6HDTy+dga26pA==",
```
Comment (char0n) [inline package-lock.json]:
```suggestion
          "integrity": "sha512-DtSLzxoiUir0qRc3+JJBxiAe6NvTEM3uDxnPxVWJU6sRDhUi8Ssx6DBjGWCZAQJlLk5A+jk2ptf3JvvZrQlLNQ==",
```

**Comment by char0n** [comment]
I've amended also changes to `package-lock.json` file. Let's wait for tests to finish and this is good to merge. Thanks for your diligent work and patience with all my comments.

**Comment by sidag95** [comment]
Thank you so much @char0n for being so patient with my work. Thanks for helping me learn without getting frustrated. I would love to contribute more and hopefully with a better PR than this one. Please let me know if I can take up any other feature request. Will keep an eye on the issues. Thanks again!

**Comment by char0n** [comment]
@sidag95 no problem at all. If you wish to continue pick any issue you like from our extensive issue list: https://github.com/char0n/ramda-adjunct/issues

**Pull request merged**